### PR TITLE
Add modern uwp (.net) support

### LINKED ISSUE #2746: [FEATURE] Support UWP for SkiaSharp 3.0
### Is your feature request related to a problem?

I am using UWP and migrating to WinUI3 is not an option (yet)

### Describe the solution you would like

I would like to use SkiaSharp 3.0 on UWP but when attempting to use it, I get the following error:
```
NU1202: Package SkiaSharp.Views 3.0.0-preview.1.8 is not compatible with uap10.0.17763 (UAP,Version=v10.0.17763) / win10-x86-aot. Package SkiaSharp.Views 3.0.0-preview.1.8 supports:
  - net462 (.NETFramework,Version=v4.6.2)
  - net7.0-android33.0 (.NETCoreApp,Version=v7.0)
  - net7.0-ios16.1 (.NETCoreApp,Version=v7.0)
  - net7.0-maccatalyst16.1 (.NETCoreApp,Version=v7.0)
  - net7.0-macos13.0 (.NETCoreApp,Version=v7.0)
  - net7.0-tizen7.0 (.NETCoreApp,Version=v7.0)
  - net7.0-tvos16.1 (.NETCoreApp,Version=v7.0)
  - net7.0-windows10.0.19041 (.NETCoreApp,Version=v7.0)
```

As announced [here](https://github.com/mono/SkiaSharp/blob/main/changelogs/SkiaSharp/3.0.0/SkiaSharp.humanreadable.md), v3.0 should be compartible with .NET Standard 2.0 (which is what UWP is using) so I am not sure if this is a bug

### Describe alternatives you have considered

Migrating to WinUI3 is not an option as it has bad performance, unsatisfying bundle size and various other issues.
I could also just stay on SkaiSharp 2.x, but I'd prefer to use the "latest and greatest" SkiaSharp version ;)

### Additional context

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by FoggyFinder**
## Description

Add modern UWP (.NET) support, including adding a sample (mostly copy of WinUI 3 version).

New packages: SkiaSharp.Views.UWP (SKXamlCanvas, SKSwapChainPanel) and SkiaSharp.NativeAssets.UWP (UWP builds of ANGLE and the WinRT helper). The WinUI view code moved to SkiaSharp.Views.WinRT.Shared and is shared by both.

No UWP-specific libSkiaSharp/libHarfBuzzSharp: the Win32 native assets work.

**Related issues**

Fixes #2746 

**Required skia PR**

None.

**Areas affected**

- [ ] Managed API (`binding/`)
- [ ] Native / C API (`externals/skia/src/c`, `include/c`)
- [ ] Generated P/Invoke bindings
- [ ] Native dependency or Skia update (libpng, HarfBuzz, FreeType, zlib, milestone bump, …)
- [x] Views & integrations (MAUI, Uno, WPF, WinUI, Blazor, …)
- [ ] Rendering output / visual behavior
- [ ] Performance
- [ ] Tests
- [x] Build, packaging, or CI
- [x] Documentation or samples

## Changes

Public API — new packages, both net10.0-windows10.0.26100.0, minimum Windows 10 1809: SkiaSharp.Views.UWP (namespace SkiaSharp.Views.UWP: SKXamlCanvas, SKSwapChainPanel, AngleSwapChainPanel, UWPExtensions, SKPaintSurfaceEventArgs, SKPaintGLSurfaceEventArgs) and SkiaSharp.NativeAssets.UWP (reuses the obsolete 2.x package ID, last 2.88.9). No existing public API changes.

Behavior — SkiaSharp.Views.WinUI: sources moved to SkiaSharp.Views.WinRT.Shared, define WINDOWS → WINUI (namespaces unchanged, Uno included). SkiaSharp.NativeAssets.WinUI: CsWinRT 2.0.4 → 2.1.6, signing runs after CsWinRTReplaceForPatchedRuntime only when SignAfterCsWinRTIIDOptimizer is set. SkiaSharp.Views (iOS, macOS, Tizen): internal GLES P/Invokes take pointers instead of out/arrays, for AOT with disabled runtime marshalling; no behaviour change intended.

## Testing

CI builds the UWP native assets, ANGLE, SkiaSharp.Views.UWP and the sample; the sample was run from Visual Studio 2026 on Windows 11 x64 (Debug and Release, all three pages); x86 and arm64 hardware untested; no automated tests, as WinUI has none either. The sample builds with MSBuild rather than the dotnet CLI, and ANGLE for UWP arm64 needs the patch file.

## Checklist

- [ ] Tests added or updated
- [x] `Changes` above lists all public API and behavioral changes
- [ ] New/changed public API?
- [ ] Native change?

### CODE REVIEW

**Comment by mattleibow** [comment]
@FoggyFinder can you have a look at the build errors.

**Comment by FoggyFinder** [comment]
> @FoggyFinder can you have a look at the build errors.

@mattleibow I wish I could but I don't have access to SkiaSharp azure. I thought I fixed all build errors today

**Comment by FoggyFinder** [comment]
seems like Copilot catched errors, I build locally as `dotnet cake --target=nuget --skipExternals=all --configuration=Release
` but probably mac/iOS is excluded by default

**Comment by FoggyFinder** [comment]
was managed to see error on a clean browser:

> Script contents:
> python3 scripts/infra/caching/repo-deps.py validate
> [FAIL] UNCOVERED FILES:
> native/uwp-angle/build.cake

I haven't looked how to updates those properly. I'll try to sort it tomorrow

File: samples/Basic/UWP/README.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+- **Color palette** — Six selectable colors with dark/light mode variants that update on theme change.
+- **Brush size** — Adjustable stroke width (1–50px) via slider or scroll wheel.
+- **Brush cursor** — Semi-transparent circle indicator showing brush size at the cursor position.
+- **Adaptive layout** — Toolbox reflows for narrow windows using `VisualStateManager` with `AdaptiveTrigger`.
+- **DPI scaling** — Canvas coordinates are scaled to handle high-DPI displays correctly.
+
+## Requirements
+
+- [.NET 10 SDK](https://dotnet.microsoft.com/download) or later
+- Windows 10 (build 19041 or later)
```
[the remark this answers is not in the document]
Comment (FoggyFinder) [reply samples/Basic/UWP/README.md]: fixed

File: source/SkiaSharp.Views/SkiaSharp.Views.WinUI_UWP/GlesInterop/GlesContext.cs [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+			Dispose(true);
+			GC.SuppressFinalize(this);
+		}
+
+		public void CreateSurface(SwapChainPanel panel, Size? renderSurfaceSize, float? resolutionScale)
+		{
+			if (panel == null)
+			{
+				throw new ArgumentNullException("SwapChainPanel parameter is invalid");
+			}
```
[the remark this answers is not in the document]
Comment (FoggyFinder) [reply source/SkiaSharp.Views/SkiaSharp.Views.WinUI_UWP/GlesInterop/GlesContext.cs:84]: unrelated to PR

File: source/SkiaSharp.Views/SkiaSharp.Views.Shared/GlesInterop/Gles.cs [hunk 3]
```diff
@@ -74,7 +73,7 @@ internal static class Gles
 		public const int GL_DEPTH24_STENCIL8_OES = 0x88F0;
 
 		[DllImport(libGLESv2)]
-		public static extern void glGetIntegerv(uint pname, out int data);
+		public static extern void glGetIntegerv(uint pname, int* data);
 		[DllImport(libGLESv2)]
```
[the remark this answers is not in the document]
Comment (FoggyFinder) [reply source/SkiaSharp.Views/SkiaSharp.Views.Shared/GlesInterop/Gles.cs:77]: thanks, fixed

**Comment by FoggyFinder** [comment]
still trying to figure out new build error for sample project

but almost out of ideas why does it happen

**Comment by FoggyFinder** [comment]
seems like dotnet build uwp limitation:

https://github.com/microsoft/microsoft-ui-xaml/discussions/9983#discussioncomment-13376111

**Comment by mattleibow** [comment]
I see there is no UWP-specific build of libSkiaSharp and libHarfBuzzSharp? Is that expected?

**Comment by FoggyFinder** [comment]
> I see there is no UWP-specific build of libSkiaSharp and libHarfBuzzSharp? Is that expected?

@mattleibow yes, works just fine as is, at least no errors I noticed.

The single error from ci on a sample project - it's because dotnet cli limitation. I tried to workaround it with dotnet restore + msbuild but it didn't work. Only newest msbuild supports uwp and sadly one from VS 2022 isn't good enough.

I don't think there is anything I can do to make it green. At least I tried many options locally and none worked.

**Comment by mattleibow** [comment]
@FoggyFinder a few things on `native/uwp-angle/build.cake`:

**1. `VER_MAJORVERSION` and `VER_MINORVERSION` are swapped in the ARM64 patch.** The shift index is the bit position of the flag (`winnt.h`):

```c
#define VER_MINORVERSION                0x0000001   // -> shift 0
#define VER_MAJORVERSION                0x0000002   // -> shift 1
```

but the reimplementation has `VER_MINORVERSION` at 1 and `VER_MAJORVERSION` at 0 (the `else` fallthrough), so every major/minor condition lands in the wrong 3-bit slot and ANGLE's `IsWindows*OrGreater()` checks would quietly return the wrong answer. The other six entries are correct.

**2. The patch may not be needed at all.** ANGLE already handles this case — `src/common/platform_helpers.cpp` has a `#if defined(ANGLE_ENABLE_WINDOWS_UWP)` branch that skips `VerifyVersionInfo` entirely and just treats UWP as Windows 10. That define comes from `config("angle_enable_winuwp_config")` (`src/libANGLE/renderer/d3d/BUILD.gn:61`), which is applied to `angle_backend_config` and `libANGLE_no_vulkan` — but `platform_helpers.cpp` lives in `angle_common`, which doesn't get it. For reference, vcpkg's ANGLE port sets `-DANGLE_ENABLE_WINDOWS_UWP -DWINAPI_FAMILY=WINAPI_FAMILY_PC_APP` together and globally via `add_definitions()`, and carries no `platform_helpers` patch.

Would you mind trying a run with the patch dropped? The guard is `#if defined(_M_ARM64)`, so it is inert for x86/x64 — and those two legs build green today, which suggests the desktop declarations are reachable in this configuration regardless. Since the patch predates the CI jobs, we have never actually built ANGLE UWP without it.

One note in case it stays: the SDK gating isn't ARM-specific. It's `WINAPI_FAMILY_PARTITION(WINAPI_PARTITION_DESKTOP | ...)` in `sysinfoapi.h`/`WinBase.h` with no architecture conditional, and both symbols are present in `WindowsApp.lib` for x86, x64 and arm64 alike — so the code comment about x86/x64 getting an inline `VerSetConditionMask` isn't quite right, and it'd be worth understanding the real ARM64-only trigger before we carry the workaround long-term. If we do keep it, could it move out into a `.patch` file applied with `git apply` rather than a string splice? That gives us proper context matching and a clear failure when it stops applying after an ANGLE bump.

**3. Could `sync-ANGLE` move into a shared helper?** It's ~90 lines duplicated almost verbatim between `native/uwp-angle/build.cake` and `native/winui-angle/build.cake` — the clone, the submodule list, the `toolchain.gni` output-name rewrite, `gclient_args.gni`, `LASTCHANGE`, the `rc.exe` download and the llvm download are all identical. Something like `scripts/infra/native/windows/angle-shared.cake` exposing `SyncAngle(anglePath, version)` would leave each build.cake with just its own tail (WinAppSDK header generation for WinUI, the patch for UWP) plus its gn args. `repo-deps.json` already covers `scripts/infra/native/windows/**` for both cache jobs, so that part needs no change.

**Comment by mattleibow** [comment]
Separate from the ANGLE comment above — a few things on the C#/packaging side:

**1. There are two copies of `Egl.cs` / `GlesContext.cs`.** `source/SkiaSharp.Views/GlesInterop/` and `source/SkiaSharp.Views/SkiaSharp.Views.WinUI_UWP/GlesInterop/` both exist. Only the `WinUI_UWP` pair is compiled (both `SkiaSharp.Views.WinUI.csproj` and `SkiaSharp.Views.UWP.csproj` glob `..\SkiaSharp.Views.WinUI_UWP\**\*.cs`); nothing references the top-level pair, and it's the unmodified original. Could the top-level copy be deleted? Then all six files read as clean renames out of `SkiaSharp.Views.WinUI/`, which makes the diff a lot easier to review.

**2. Could the shared folder be named `SkiaSharp.Views.WinRT.Shared`?** `WinUI_UWP` doesn't match the repo convention — the existing source-only folders are `SkiaSharp.Views.Shared` and `SkiaSharp.Views.Uno.WinUI.Shared`, and every folder under `source/SkiaSharp.Views/` without a `.Shared` suffix is a real project, so a bare `SkiaSharp.Views.WinUI_UWP` sitting next to `SkiaSharp.Views.WinUI` reads like a project that lost its csproj. WinRT is the accurate common denominator (UWP is `Windows.UI.Xaml`, WinUI 3 is `Microsoft.UI.Xaml`, both WinRT + CsWinRT). It's five `<Compile Include>` paths to update — the two views projects plus the three Uno ones that pull in `UWPExtensions.cs`.

**3. `SkiaSharp.Views.UWP` needs a different assembly name.** It currently sets

```xml
<RootNamespace>SkiaSharp.Views.Windows</RootNamespace>
<AssemblyName>SkiaSharp.Views.Windows</AssemblyName>
```

which is byte-identical to `SkiaSharp.Views.WinUI`, but the types inside are in `SkiaSharp.Views.UWP` (the `#else` branch of the namespace switch), and the sample already does `using SkiaSharp.Views.UWP;`. Two packages shipping `SkiaSharp.Views.Windows.dll` with different public surfaces will collide — including in the docs build, where both map to the `skiasharp-views` moniker and the staging logic keeps whichever assembly has more types, so one package's API would silently vanish. `SkiaSharp.Views.UWP` for both properties would match the namespace, the package id and the old package.

**4. Could the CsWinRT versions be unified on `2.3.1`?** Right now there are three different ones:

| project | version |
|---|---|
| `native/uwp/…UWP.Native.Projection` | `2.3.0-prerelease.251115.2` |
| `source/…/SkiaSharp.Views.UWP` | `2.3.0-prerelease.251115.2` |
| `native/winui/…WinUI.Native.Projection` | `2.1.0` (bumped from `2.0.4` in this PR) |

`2.3.1` is the stable release of that 2.3 line, so it supersedes both prereleases and we'd avoid shipping a prerelease dependency in a public package. It also pairs with `WindowsSdkPackageVersion` `10.0.26100.87` (the ref pack CsWinRT 2.3.1 ships for .NET 9/10) rather than the current `10.0.26100.57`. Worth noting `SkiaSharp.Views.UWP.Native.Projection.csproj` doesn't set `WindowsSdkPackageVersion` at all, so it inherits the `10.0.19041.38` pinned in `SkiaSharp.Build.props` for WinUI while targeting 26100 — might be cleaner to put the UWP value in `SkiaSharp.Build.props` next to `$(UWPTargetFramework)` and have both UWP projects reference it.

**5. `native/uwp/build.cake` doesn't copy the `.winmd`.** The vcxproj sets `<GenerateWindowsMetadata>true</GenerateWindowsMetadata>` and uses the same `OutDir` as WinUI, so it is being produced — it's just not copied to the output directory the way `native/winui/build.cake` does:

```csharp
CopyFileToDirectory($"{name}/bin/{nativeArch}/{CONFIGURATION}/{name}.winmd", outDir);
```

**Comment by FoggyFinder** [comment]
1. fixed
2. done
3. fixed, it's me overthinking. I thought people may confuse old and modern uwp :D
4. stable package is finally released! I tried to bump version to `2.3.1`/`10.0.26100.87` but then `_SignAssemblyVerify` target fails. Not sure why yet
5. done. At first I thought winmd is required for uwp as well but apparantly not.

I'll try to address angle-related comment next.
There was build error which was a complete mystery to me for arm64 so I let Claude fix it - since build was green I didn't check patch and comment carefully. That's my bad.

**Comment by mattleibow** [comment]
no problem, that is why I got my copilot to review and judge your claude.

**Comment by FoggyFinder** [comment]
`Microsoft.Windows.CsWinRT` is updated to 2.3.1 for uwp, it's not compatible with current sdk (`10.0.19041.38`) for winui so I left it as is

> a few things on native/uwp-angle/build.cake:

@mattleibow 

1. fixed
2. patch was added because ninja failed with

```
../../../src/common/platform_helpers.cpp(46): error C3861: VerSetConditionMask: identifier not found
../../../src/common/platform_helpers.cpp(47): error C3861: VerSetConditionMask: identifier not found
../../../src/common/platform_helpers.cpp(48): error C3861: VerSetConditionMask: identifier not found
../../../src/common/platform_helpers.cpp(56): error C3861: VerifyVersionInfoW: identifier not found
```

for arm64.

Claude isn't helpful here - contradicts itself with different version and I'm not qualified to verify its claims. As long as it builds locally any explanation sounds plausible to me. Maybe it's something to look into carefully, not sure.

Never used `git apply` before, I'll try to check what it is

3. done

**Comment by charlesroddie** [comment]
@mattleibow do you know what more this needs to go in?

File: source/SkiaSharp.Views/SkiaSharp.Views.WinUI_UWP/GlesInterop/GlesContext.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (charlesroddie) [reply source/SkiaSharp.Views/SkiaSharp.Views.WinUI_UWP/GlesInterop/GlesContext.cs:84]: Fixed in https://github.com/mono/SkiaSharp/pull/4540/commits/b13862ad9985982ec55a124a2ec1d8c9832e149f

**Comment by charlesroddie** [comment]
Updated this branch with some fixes, and CI is working now.
Tested on Windows 11 x64.

Changes reported by Claude:

```
- **Merged `main`** and resolved conflicts (`.gitattributes`, `VERSIONS.txt`, Windows solution filter).
- **Copilot's review-summary comments:** restored null propagation in `SKPicture.ToWriteableBitmap`, removed the redundant wait in the UWP `RenderLoop`, fixed the `AddSize` parameter name and the UWP native readme.
- **Registration:** UWP binaries added to `eng/Signing.props`; `SkiaSharp.NativeAssets.UWP` moved from obsoleted to supported NuGets and `SkiaSharp.Views.UWP` added (tracking version 4.154.0); `packages.md` and the Views package README updated.
- **API Scan:** ARM64 surrogates for the UWP `libEGL`, `libGLESv2` and `SkiaSharp.Views.UWP.Native`, copied to `surrogate-binaries/uwp` to avoid name collisions with WinUI. Test updated.
- **CI fixes:** UWP native jobs now install `DOTNET_VERSION_WINUI` like WinUI (they failed to resolve the SDK after the merge), and the UWP sample restore uses the local NuGet sources.
- **Sample README:** minimum OS is 1809 (17763), and it needs Visual Studio 2026 to run.
```

**Comment by charlesroddie** [comment]
Last two commits ensure that Uno keeps the UWPExtensions name, and document build symbols (which in the past I found one of the hardest tings about making changes in this repo).

Local claude opus 5 review is clean.

@mattleibow need anything else?
